src/materials/Material.d.ts

```diff
@@ -139,7 +139,7 @@ export class Material extends EventDispatcher {
 	 * The pairs are defined in both vertex and fragment shaders. Default is undefined.
 	 * @default undefined
 	 */
-	defines: { [key: string]: any };
+	defines: undefined | { [key: string]: any };
```

Comment:
This should be safe since all materials which actually set defines like `MeshStandardMaterial` overwrite the type declaration in their own TS file with:
```js
defines: { [key: string]: any };
```
So property access in context of those materials should not break.

Reply:
It would be better modeling the type exactly for that .js unless the item is internal.
for this case Material is exported (not internal), one can `class MyMaterial extends Material {}` in TS.

```ts // ts
this.defines['x'] = 0; // passed typecheck, fail at runtime.
```

Reply:
How about defining `defines` in `Material` like so:
```js
this.defines = {};
```
In this way, it would be never `undefined`. An we could delete all type declarations of `defines` in derived materials (e.g. `ShaderMaterial`).

Reply:
> 
> 
> How about defining `defines` in `Material.js` like so:
> 
> ```js
> this.defines = {};
> ```
> 
> In this way, it would be never `undefined`. An we could delete all type declarations of `defines` in derived materials (e.g. `ShaderMaterial`).

This will make those derived materials (point, sprite, shadow..) bloated.

Reply:
What do you mean with "bloated"? Do you mind explaining in more detail?

Reply:
> 
> 
> What do you mean with "bloated"? Do you mind explaining in more detail?


After adding `.defines = {}` to Material.js; it is fine from shadermaterial POV as it needs this property.
But for those derived materials like pointsmaterial, spritematerial, shadowmaterial etc, 
they own a `.defines` (an empty js object) which is never be used.

Reply:
But this also true for other common material properties. E.g. `opacity` is not use by all materials.